# SocialNetworkAnalysis Final Project

# Abstract
This study explores the in silico effects of miR-302b treatment on mouse liver cells. Starting from the differentially expressed genes identified by the original researchers, the raw RNA-seq data was re-analyzed using the DESeq2 algorithm to obtain gene-level fold changes. A significance threshold was then applied to select a subset of highly differentially expressed genes, which served as "seed" nodes for constructing a protein–protein interaction (PPI) network via the STRING database API. The extended network, including first- and second-order interactors, comprised 2070 additional genes not present in the initial dataset. The analysis was divided into three main sections. The first involved characterizing the topological properties of the network—such as degree distribution, betweenness centrality, and diameter—and comparing them with Erdős–Rényi and Barabási–Albert models. The second section consisted in performing community detection and evaluating the resulting modules through gene set enrichment analysis (GSEA). Finally, the potential of fold-changes network propagation on a signed, directed graph was investigated to infer possible new target genes that might be influenced by the treatment. Collectively, these analyses provide an integrated systems-level perspective on the molecular impact of miR-302b treatment in hepatic cells.
